src/Razor/src/Microsoft.AspNetCore.Razor.LanguageServer/AutoInsert/AutoClosingTagOnAutoInsertProvider.cs

```diff
@@ -0,0 +1,201 @@
+                startTag.Parent is MarkupElementSyntax)
+            {
+                name = startTag.Name.Content;
+                autoClosingBehavior = InferAutoClosingBehavior(name);
+                name = startTag.GetTagNameWithOptionalBang();
+                return true;
+            }
+
+            if (owner.Parent is MarkupTagHelperStartTagSyntax startTagHelper &&
+                startTagHelper.Parent is MarkupTagHelperElementSyntax tagHelperElement)
+            {
+                name = startTagHelper.Name.Content;
+
+                if (!TryGetTagHelperAutoClosingBehavior(tagHelperElement.TagHelperInfo.BindingResult, out autoClosingBehavior))
+                {
+                    autoClosingBehavior = InferAutoClosingBehavior(name);
+                }
+
+                return true;
+            }
+
+            autoClosingBehavior = default;
+            name = null;
+            return false;
+        }
+
+        private static AutoClosingBehavior InferAutoClosingBehavior(string name)
+        {
+            if (s_VoidElements.Contains(name))
+            {
+                return AutoClosingBehavior.SelfClosing;
+            }
+
+            return AutoClosingBehavior.Default;
+        }
+
+        private static bool TryGetTagHelperAutoClosingBehavior(TagHelperBinding bindingResult, out AutoClosingBehavior autoClosingBehavior)
+        {
+            var resolvedTagStructure = TagStructure.Unspecified;
+
+            foreach (var descriptor in bindingResult.Descriptors)
+            {
+                var tagMatchingRules = bindingResult.GetBoundRules(descriptor);
+                for (var i = 0; i < tagMatchingRules.Count; i++)
+                {
+                    var tagMatchingRule = tagMatchingRules[i];
+                    if (tagMatchingRule.TagStructure == TagStructure.NormalOrSelfClosing)
+                    {
+                        // We have a rule that indicates it can be normal or self-closing, that always wins because.
+                        // it's all encompassing. Meaning, even if all previous rules indicate "no children" and at least
+                        // one says it supports children we render the tag as having the potential to have children.
+                        autoClosingBehavior = AutoClosingBehavior.Default;
+                        return true;
+                    }
+
+                    resolvedTagStructure = tagMatchingRule.TagStructure;
+                }
+            }
+
+            Debug.Assert(resolvedTagStructure != TagStructure.NormalOrSelfClosing, "Normal tag structure should already have been preferred");
+
+            if (resolvedTagStructure == TagStructure.WithoutEndTag)
+            {
+                autoClosingBehavior = AutoClosingBehavior.SelfClosing;
+                return true;
+            }
+
+            autoClosingBehavior = default;
+            return false;
+        }
+
+        private enum AutoClosingBehavior
+        {
+            Default,
```

Comment:
Having the options here be `SelfClosing` and `Default` feels weird, though I'm not sure what you'd call the other behavior (EndTag?).

Reply:
Kind of like `EndTag`, i'll change